### PR TITLE
Improve reassembly worst-case performance

### LINKED ISSUE #575: Reassembly worst-case performance
This pcap shows extremely poor TCP reassembly performance:

[pathological-reassembly-duplicate-segments.pcap.gz](https://github.com/zeek/zeek/files/3612111/pathological-reassembly-duplicate-segments.pcap.gz)

We stop seeing ACKs from one endpoint while its peers sends a duplicates/retransmissions for every packet.  Probably an unlikely case (I artificially made the pcap by duplicating packets), but the reassembly overlap checking becomes O(n) for every packet as the number of buffered packets keeps growing (so it becomes ~50k iterations over a linked list for each packet).

### PR DESCRIPTION
**Pull request by jsiwek**
Fixes #575

Changes the internal reassembly data structure to use `std::map` instead of a linked-list.

# Performance Comparisons Versus `master`

## Common Case
Timings using `2009-M57-day11-18.trace`

- branch / debug build / bare scripts: 3.102
- branch / debug build / full scripts: 10.498
- branch / release build / bare scripts: 0.946
- branch / release build / full scripts: 4.717
- master / debug build / bare scripts: 2.84
- master / debug build / full scripts: 10.125
- master / release build / bare scripts: 0.96
- master / release build / full scripts: 4.664

Additional overhead is noticeable for debug build.  Within 1% overhead for release build.

## Pathological Case

Using `pathological-reassembly-duplicate-segments.pcap`

Master:
```
$ time -f "%e" zeek -b -r pathological-reassembly-duplicate-segments.pcap
15.77
```

Branch:

```
$ time -f "%e" zeek -b -r pathological-reassembly-duplicate-segments.pcap
0.21
```

Way better.

# Questions

## Would `std::vector` perform better than `std::map`

E.g. due to better cache use.

Maybe, I tested it and was overall like 1% better than `std::map` in debug
build, but that's also only testing the common case where 99% if insertions were
at the end of the list.  Safer to just use a map.

## Would a skip-list perform better than `std::map`

Seems like it. Tested in `topic/jsiwek/reassembly-improvements` and could not
tell any difference from the old linked-list implementation using the most naive
skip-list implementation.  Inserting and erasing a node in a tree just always
adds a bit more work.

Reasons not to (initially) use a skip-list:

- Overhead of `std::map` seems minimal, so better to spend our time
  working on other areas

- Skip-list isn't deterministic, so technically it can randomly, with low
  probability, still hit the O(n) worst case.  There's ways to make it
  deterministic, but didn't fully investigate -- initial understanding is that it
  requires periodically iterating over the full list to balance it, so may 
  end up with similar overhead to `std::map` in the end.

- Additional complexity.  Harder to audit correctness.  Additional maintenance
  burden.

## Should we add an option similar to `tcp_excessive_data_without_further_acks`, but instead measured by number of blocks ?

Maybe, but I don't know what number to pick as the default.  The problem is that
we release the reassembly buffer based on ACKs, so it's effected by window size
-- large window, small segments means it's maybe easy to pick a bad default and
skip analyzing more legit connections than we want.

### CODE REVIEW

**Comment by jsiwek** [comment]
Clarification to the `std::vector` performance: was using an ordered vector for looking up blocks by a sequence number, but data from blocks themselves all still allocated separately in possibly non-contiguous memory locations (same as it is now).

**Comment by jsiwek** [comment]
@rsmmr had the idea of dynamically changing between a `vector` (or `list`) and a `map` depending on number of blocks being buffered (help minimize the overhead of `map` for more common cases of smaller quantity of blocks in reassembly buffers).

I think that's a neat idea to try, but now realize it's a bit involved to adapt things to get that working -- quite a bit of logic revolves around iterators and those end up looking different for `vector` vs. `map`.  So we'd want a new container class that wraps the details of "use a `vector` if below N elements, else use `map`", but even then, the iterator semantics get a bit weird: insertions in a `vector` may invalidate iterators, but insertions to a `map` never do.  We do have at least one code path that currently relies on insertions not causing an iterator invalidation. Can likely rewrite such spots to deal with the sudden change from `vector` to `map`, but the additional complexity makes me want to punt on getting that experiment working unless we're really bent on getting back our 1% `map` overhead.

There's also additional uncertainty of what to choose for the cutoff -- pick something too low and we'll just end up wasting time switching over most things to use a `map` anyway.  Pick something too high and then we potentially still see some cases where reassembly performs poorly (slow iteration in case of `list` or large resizing operations for case of `vector`).

TLDR: still most comfortable with the plain `map` implementation for now

**Comment by rsmmr** [comment]
Left two small comments but will go ahead and merge, as these are crucial.

File: src/Reassem.cc [hunk 1]
```diff
@@ -1,354 +1,372 @@
...
+DataBlock::DataBlock(const u_char* data, uint64_t size, uint64_t arg_seq)
 	{
 	seq = arg_seq;
 	upper = seq + size;
 	block = new u_char[size];
+	memcpy(block, data, size);
+	}
 
-	memcpy((void*) block, (const void*) data, size);
+void DataBlockList::DataSize(uint64_t seq_cutoff, uint64_t* below, uint64_t* above) const
```
[the remark this answers is not in the document]
Comment (jsiwek) [reply src/Reassem.cc]: Yeah, thanks.  Went ahead to just change the method to check for a cutoff in the middle of a block.

File: src/Reassem.h [hunk 2]
```diff
@@ -20,30 +22,235 @@ enum ReassemblerType {
...
+	 * Must not be called when the list is empty.
+	 */
+	const DataBlock& FirstBlock() const
+		{ return block_map.begin()->second; }
+
+	/**
+	 * @return reference to the last data block in the list.
+	 * Must not be called when the list is empty.
+	 */
+	const DataBlock& LastBlock() const
```
[the remark this answers is not in the document]
Comment (jsiwek) [reply src/Reassem.h]: The cumbersome part is mostly the repetition of `std::prev(block_list.End())->second`, which I replaced with `block_list.LastBlock()`, but I prefer the explicit logic to check for "empty list" rather than indirectly via "null last block".  Not really opposed to adding a new method for that, but just going to leave it up to whoever else next works on the code to add if it's useful to them.
